src/Featurizers/AnalyticalRollingWindowFeaturizer.h

```diff
@@ -0,0 +1,307 @@
+
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+    bool begin_training_impl(void) override;
+
+    // Note that like in the InferenceOnlyEstimator the following training methods aren't used, but need to be overridden as
+    // the base implementations are abstract. The noop definitions are below.
+    FitResult fit_impl(InputT const *pBuffer, size_t cBuffer) override;
+
+    void complete_training_impl(void) override;
+
+    // MSVC has problems when the definition for the func is separated from its declaration.
+    typename BaseType::TransformerUniquePtr create_transformer_impl(void) override {
+        return typename BaseType::TransformerUniquePtr(new AnalyticalRollingWindowTransformer<InputT, MaxNumTrainingItemsV>(_windowSize, _windowCalculation, _horizon, _minWindowCount));
+    }
+
+};
+
+/////////////////////////////////////////////////////////////////////////
+///  \typedef       GrainedAnalyticalRollingWindowEstimator
+///  \brief         GrainedTransformer that creates `AnalyticalRollingWindowEstimator`.
+///
+// Normally we want the featurizer name at the top of the file, but GrainFeaturizers are just type defs, so we want the name next to it.
+static constexpr char const * const         GrainedAnalyticalRollingWindowEstimatorName("GrainAnalyticalRollingWindowEstimator");
+
+template <typename InputT, size_t MaxNumTrainingItemsV=std::numeric_limits<size_t>::max()>
+using GrainedAnalyticalRollingWindowEstimator = Components::GrainEstimatorImpl<std::vector<std::string>, AnalyticalRollingWindowEstimator<InputT, MaxNumTrainingItemsV>>;
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// |
+// |  Implementation
+// |
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+
+// ----------------------------------------------------------------------
+// |
+// |  AnalyticalRollingWindowTransformer
+// |
+// ----------------------------------------------------------------------
+template <typename InputT, size_t MaxNumTrainingItemsV>
+AnalyticalRollingWindowTransformer<InputT, MaxNumTrainingItemsV>::AnalyticalRollingWindowTransformer(std::uint32_t windowSize, AnalyticalRollingWindowCalculation windowCalculation, std::uint32_t horizon, std::uint32_t minWindowCount) :
+    _windowSize(std::move(windowSize)),
+    _windowCalculation(std::move(windowCalculation)),
+    _horizon(std::move(horizon)),
+    _minWindowCount(std::move(minWindowCount)),
+    _buffer(horizon + (windowSize)) {
+

```

Comment:
Validate these values before using them.

Reply:
With something like this is it best to validate in the Estimator, the Transformer, or both? I think just the Estimator would work. Anything I am missing with that assumption?

Reply:
Per our offline conversation: both.

---
In reply to: [395305722](https://github.com/microsoft/FeaturizersLibrary/pull/173#discussion_r395305722) [](ancestors = 395305722)

Reply:
Validation has been added.